### ISSUE TITLE
Learning Rate Finder doesn't reset model parameters

### ISSUE THREAD

**Issue by toliz** [PyTorchLightning/pytorch-lightning#6675]
## 🐛 Bug

LR finder does not reset model state after running. For big enough values of `max_lr`  for the model to diverge, if training starts right after lr tuning, model diverges.

This means the user has to manually reset the model parameters every time after tuning model's learning rate.

### To Reproduce

Use the following [**BoringModel**](https://colab.research.google.com/drive/1HvWVVTK8j2Nj52qU4Q4YCyzOm0_aLQF3?usp=sharing) and post here

### Expected behavior

Learning Rate Finder should reset models state. Maybe call module's `__init__()`?

### Environment

* CUDA:
	- GPU:
		- GeForce GTX 1080
	- available:         True
	- version:           10.2
* Packages:
	- numpy:             1.19.2
	- pyTorch_debug:     False
	- pyTorch_version:   1.6.0
	- pytorch-lightning: 1.2.1
	- tqdm:              4.55.1
* System:
	- OS:                Linux
	- architecture:
		- 64bit
		- ELF
	- processor:         x86_64
	- python:            3.8.6
	- version:           #140-Ubuntu SMP Thu Jan 28 05:20:47 UTC 2021

**Comment by justusschock** [comment]
@toliz Thanks for your issue. You're right, we don't do that and this is kind-of on purpose. LightningModules often don't only contain the model but are systems and reinitializing them could be time consuming, which is why we don't do this. However, you are right, that some kind of logic for resetting the weights may be necessary. What do you think about a specific hook for that, that can be implemented?

@SkafteNicki could we per default create a deepcopy of the weights before tuning (on cpu) and load them afterwards?

**Comment by SkafteNicki** [comment]
I do not understand this. We explicitly dump the model state to a file before the learning rate finder
https://github.com/PyTorchLightning/pytorch-lightning/blob/9be092dbdb37db353002365e6d5219fae116cc42/pytorch_lightning/tuner/lr_finder.py#L173
and then load afterwards to restore the model state
https://github.com/PyTorchLightning/pytorch-lightning/blob/9be092dbdb37db353002365e6d5219fae116cc42/pytorch_lightning/tuner/lr_finder.py#L191
should this not reset the weights?

**Comment by toliz** [comment]
@justusschock I understand. Definitely not reinitialize the lightning module then.

I check the code @SkafteNicki mentioned and from what I see there has been taken care of storing and loading model weights. Maybe there is a bug then?

Also the [doctoring](https://github.com/PyTorchLightning/pytorch-lightning/blob/9be092dbdb37db353002365e6d5219fae116cc42/pytorch_lightning/tuner/lr_finder.py#L129-L131) of `lr_find` motivates the user to redefine the model (which as you said can be time consuming)

**Comment by tchaton** [comment]
Dear @toliz,

Just checked and we do have a test to assert the model state is properly reset.

```
def test_model_reset_correctly(tmpdir):
    """ Check that model weights are correctly reset after lr_find() """

    model = EvalModelTemplate()

    # logger file to get meta
    trainer = Trainer(
        default_root_dir=tmpdir,
        max_epochs=1,
    )

    before_state_dict = deepcopy(model.state_dict())

    _ = trainer.tuner.lr_find(model, num_training=5)

    after_state_dict = model.state_dict()

    for key in before_state_dict.keys():
        assert torch.all(torch.eq(before_state_dict[key], after_state_dict[key])), \
            'Model was not reset correctly after learning rate finder'

    assert not os.path.exists(tmpdir / 'lr_find_temp_model.ckpt')
```

Could it be that the lr_find failed and the value being too high your model diverge ?

Best,
T.C

**Comment by Darshan-Ramesh** [comment]
@toliz I can also confirm the same behavior. The auto_lr_finder is broken or has some weird bug. My loss also starts off with a huge number and the model doesn't converge at all using the learning rate from the auto_lr_finder. But with a manually set learning rate, my model has an excellent validation f1 score.

**Comment by akihironitta** [comment]
> The auto_lr_finder is broken or has some weird bug.

@toliz @Darshan-Ramesh In some cases (specifically, when the loss diverges), lr_finder may suggest a too high learning rate as reported in #6471. The fix is now under review in #7076.

I'm still unsure why "Learning Rate Finder doesn't reset model parameters", but the bug above may have caused the behaviour you mentioned, so could you wait for the fix to get included in the next release or directly try the branch in #7076?

**Comment by akihironitta** [comment]
@toliz @Darshan-Ramesh Now that the fix for lr_finder suggesting a too high learning rate has been merged, could you try with the latest version? If the problem still persists, could you share your script to reproduce the behaviour?

**Comment by Darshan-Ramesh** [comment]
@akihironitta Thanks for the fix 😀 
For my project, with the same settings as before and with the latest version (1.3.0), I see that there is no problem as before. My model is learning properly from the automatically found learning rate.

**Comment by grudloff** [comment]
I am facing a similar issue, not sure if related. But if `lr_find` is called multiple times without reinstantiating the trainer, the behavior is the same as not resetting the weights of the model. Something like this would lead to that behaviour:

```python
#init model
#init optimizer
#init trainer
for n in [100, 200, 300]:
    #trainer should be instantiated here for correct behaviour
    lr_finder = trainer.tuner.lr_find(model, ... , num_training=n)
    # do something with lr_finder
```

**Comment by tchaton** [comment]
Hey @grudloff,

Yes, this is expected. Calling multiple time the trainer with lr_find is quite a weird behaviour. Which weights should be used to optimize the model.

I think creating a new Trainer is definitely the safest solution there.

Best,
T.C

**Comment by grudloff** [comment]
Hello @tchaton, thanks for the response. 

I am working on a dataset that allows very large batches, and as a consequence, few of them, because of this I had to find a heuristic for an appropriate number of num_training. Indeed I haven't found any information of anyone having to do something like that.

Yes, the solution works fine, but I am still puzzled, why would this be the expected behavior, and why does it behave like that? lr_find is supposed to roll back the weights when finished, how is the Trainer maintaining a reference to the 'trained' weights?

**Comment by tchaton** [comment]
Oh I think I might have over-looked your use-case. Yes, this could be a bug.

Would you mind creating a reproducible script with the BoringModel ?
